File: ck_multiple linear regression.py
# -*- coding: utf-8 -*-
"""
Created on Tue Nov 13 13:10:43 2018

@author: dkali
"""
# Multiple linear regression (mlr)

# Data Preprocessing
# Importing the libraries
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd
# Importing the dataset
dataset = pd.read_csv('50_Startups.csv')
X = dataset.iloc[:, :-1].values
y = dataset.iloc[:, -1].values
# Encoding categorical data
# Encoding the Independent Variable
from sklearn.preprocessing import LabelEncoder, OneHotEncoder
labelencoder_X = LabelEncoder()
X[:, 3] = labelencoder_X.fit_transform(X[:, 3])
onehotencoder = OneHotEncoder(categorical_features = [3])
X = onehotencoder.fit_transform(X).toarray()

# Avoiding the dummy variable trap
 X = X[:, 1:]
 
# Splitting the dataset into the Training set and Test set
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = .2, random_state = 0)
#..........................................
# Fitting mlr to train sets
from sklearn.linear_model import LinearRegression
regressor = LinearRegression()
regressor.fit(X_train, y_train) 

# Predicting the test set result
y_pred = regressor.predict(X_test)

# Building the optimal model using the Backward elimination
import statsmodels.formula.api as sm # does not include b0 in model build
# add x0 = 1s to X to cater for missing b0 in statsmodel lib.
X = np.append(arr = np.ones((50, 1)).astype(int), values = X, axis = 1 )
X_opt = X[:, [0,1,2,3,4,5]]
regressor_OLS = sm.OLS(endog = y, exog = X_opt).fit() 
regressor_OLS.summary()

X_opt = X[:, [0,1,3,4,5]]
regressor_OLS = sm.OLS(endog = y, exog = X_opt).fit() 
regressor_OLS.summary()
X_opt = X[:, [0,3,4,5]]
regressor_OLS = sm.OLS(endog = y, exog = X_opt).fit() 
regressor_OLS.summary()
X_opt = X[:, [0,3,5]]
regressor_OLS = sm.OLS(endog = y, exog = X_opt).fit() 
regressor_OLS.summary()
X_opt = X[:, [0,3]]
regressor_OLS = sm.OLS(endog = y, exog = X_opt).fit() 
regressor_OLS.summary()
X[0]
len(X[0])
# Automated backward elimination using p-value only
import statsmodels.formula.api as sm
def backwardElimination(x, sl):
    numVars = len(x[0])
    for i in range(0, numVars):
        regressor_OLS = sm.OLS(y, x).fit()
        maxVar = max(regressor_OLS.pvalues).astype(float)
        if maxVar > sl:
            for j in range(0, numVars - i):
                if (regressor_OLS.pvalues[j].astype(float) == maxVar):
                    x = np.delete(x, j, 1)
    regressor_OLS.summary()
    return x
 
SL = 0.05
X_opt = X[:, [0, 1, 2, 3, 4, 5]]
X_Modeled = backwardElimination(X_opt, SL)

File: ck_multiple linear regression.R
# Multiple linear regression
## Data Preprocessing Template
# Importing the dataset
dataset = read.csv('50_Startups.csv')
# Encoding categorical data
dataset$State <- factor(dataset$State,
                        levels = c('California', 'Florida', 'New York'),
                        labels = c(1, 2, 3) )
# Splitting the dataset into the Training set and Test set
# install.packages('caTools')
library(caTools)
set.seed(123)
split = sample.split(dataset$Profit, SplitRatio = .8)
training_set = subset(dataset, split == TRUE)
test_set = subset(dataset, split == FALSE)
#.................................................................
# Fitting mlr to the training sets
regressor = lm(formula = Profit ~ . , data = training_set)
# Predicting test set values
y_pred = predict(regressor, newdata = test_set)
# just seeing if sig. variable is a better estimator
reg1 = lm(formula = Profit ~ R.D.Spend, data = training_set)
ypred1 = predict(reg1, newdata = test_set)

# Building the optimal model using backward elimination
regressor = lm(formula = Profit ~ R.D.Spend + Administration + Marketing.Spend + State,
               data = dataset)
summary(regressor)
regressor = lm(formula = Profit ~ R.D.Spend + Administration + Marketing.Spend,
               data = dataset)
summary(regressor)
regressor = lm(formula = Profit ~ R.D.Spend + Marketing.Spend,
               data = dataset)
summary(regressor)
regressor = lm(formula = Profit ~ R.D.Spend,
               data = dataset)
summary(regressor)
# Automated backward elimination model based on p-value
backwardElimination <- function(x, sl) {
        numVars = length(x)
        for (i in c(1:numVars)){
                regressor = lm(formula = Profit ~ ., data = x)
                maxVar = max(coef(summary(regressor))[c(2:numVars), "Pr(>|t|)"])
                if (maxVar > sl){
                        j = which(coef(summary(regressor))[c(2:numVars), "Pr(>|t|)"] == maxVar)
                        x = x[, -j]
                }
                numVars = numVars - 1
        }
        return(summary(regressor))
}

SL = 0.05
dataset = dataset[, c(1,2,3,4,5)]
backwardElimination(training_set, SL)